### PR TITLE
[WIP] Add a CommandWithSubcommands class

### PR DESCRIPTION
**Pull request by phlax**
for use by pootle_fs subcommands

### CODE REVIEW

**Comment by phlax** [comment]
heres some example output from a command i was testing with

```
(pootle)phlax@aio:~/.virtualenvs/pootle/pootle$ pootle sub_test -h
2016-06-08 12:12:30,362 INFO Loading custom settings from '/home/phlax/.pootle/pootle.conf'...
usage: pootle sub_test [-h] [--version] [-v {0,1,2,3}] [--settings SETTINGS]
                       [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                       {foo,bar} ...

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -v {0,1,2,3}, --verbosity {0,1,2,3}
                        Verbosity level; 0=minimal output, 1=normal output,
                        2=verbose output, 3=very verbose output
  --settings SETTINGS   The Python path to a settings module, e.g.
                        "myproject.settings.main". If this isn't provided, the
                        DJANGO_SETTINGS_MODULE environment variable will be
                        used.
  --pythonpath PYTHONPATH
                        A directory to add to the Python path, e.g.
                        "/home/djangoprojects/myproject".
  --traceback           Raise on CommandError exceptions
  --no-color            Don't colorize the command output.

subcommands:
  {foo,bar}
    bar                 Bar sub command has help!
(pootle)phlax@aio:~/.virtualenvs/pootle/pootle$ pootle sub_test -h bar
2016-06-08 12:12:34,078 INFO Loading custom settings from '/home/phlax/.pootle/pootle.conf'...
usage: pootle sub_test [-h] [--version] [-v {0,1,2,3}] [--settings SETTINGS]
                       [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                       {foo,bar} ...

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  -v {0,1,2,3}, --verbosity {0,1,2,3}
                        Verbosity level; 0=minimal output, 1=normal output,
                        2=verbose output, 3=very verbose output
  --settings SETTINGS   The Python path to a settings module, e.g.
                        "myproject.settings.main". If this isn't provided, the
                        DJANGO_SETTINGS_MODULE environment variable will be
                        used.
  --pythonpath PYTHONPATH
                        A directory to add to the Python path, e.g.
                        "/home/djangoprojects/myproject".
  --traceback           Raise on CommandError exceptions
  --no-color            Don't colorize the command output.

subcommands:
  {foo,bar}
    bar                 Bar sub command has help!

```

File: pootle/core/management/subcommands.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            known = None
+        is_subcommand = (
+            known
+            and hasattr(known, "subcommand")
+            and known.subcommand in self.cmd.subcommands)
+        if is_subcommand:
+            # if there is a subcommand then move it in argv to the front of other
+            # params and defer to the the subcommand
+            args = list(args)
+            del args[args.index(known.subcommand)]
```
Comment (unho) [inline pootle/core/management/subcommands.py]: If the following is used we can avoid converting to `list` and using `del`:

```python
            args = tuple(item for item in args if item != known.subcommand)
```

Comment (unho) [inline pootle/core/management/subcommands.py]: I am now thinking that this won't remove the subcommand from the previous `args` object which might be still used in other places. So forget about my proposal.

File: pootle/core/management/subcommands.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        # description
+        formatter.add_text(self.description)
+
+        # positionals, optionals and user-defined groups for parent
+        for action_group in parent._action_groups:
+            # dont include the subcommands from parent
+            is_subcommands = (
+                action_group._group_actions
+
```
Comment (unho) [inline pootle/core/management/subcommands.py]: Superfluous blank line.

File: pootle/core/management/subcommands.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                and action_group._group_actions[0].dest == "subcommand")
+            if is_subcommands:
```
Comment (unho) [inline pootle/core/management/subcommands.py]: `is_subcommands`? `is_subcommand` perhaps?

File: pootle/core/management/subcommands.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+from pootle.core.delegate import subcommands
+
+
+class SubcommandsError(CommandError):
+
+    pass
+
+
+class SubcommandsParser(CommandParser):
```
Comment (unho) [inline pootle/core/management/subcommands.py]: Does it parse more than one subcommand?

File: pootle/core/management/subcommands.py (hunk 3 above)
Comment (phlax) [inline pootle/core/management/subcommands.py]: no, its a `Subparser*s*` object

File: pootle/core/management/subcommands.py (hunk 4 above)
Comment (phlax) [inline pootle/core/management/subcommands.py]: yes, correct

File: pootle/core/management/subcommands.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        Override django's run_from_argv to parse a subcommand.
+        If the subcommand is present, then mangle argv and defer to the
+        subcommand
+        """
+        self._called_from_command_line = True
+        parser = self.create_parser(argv[0], argv[1])
+
+        try:
+            # first find if we have a known subcommand
+            known, __ = parser.parse_known_args(argv[2:])
```
Comment (unho) [inline pootle/core/management/subcommands.py]: Why two underscores?

File: pootle/core/management/subcommands.py (hunk 4 above)
Comment (unho) [inline pootle/core/management/subcommands.py]: Then nothing to raise.

File: pootle/core/management/subcommands.py (hunk 3 above)
Comment (unho) [inline pootle/core/management/subcommands.py]: Then `is_subcommands` is a bit misleading, isn't it?

File: pootle/core/management/subcommands.py (hunk 5 above)
Comment (phlax) [inline pootle/core/management/subcommands.py]: so as not to conflict with `_` but to make it throwaway

**Comment by unho** [comment]
Nits aside, lgtm.

File: tests/core/management/subcommands.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    opt_mapping = {
+        sorted(s_opt.option_strings)[0].lstrip('-').replace('-', '_'): s_opt.dest
+        for s_opt in parser._actions if s_opt.option_strings}
+    arg_options = {
+        opt_mapping.get(key, key): value for key, value in options.items()}
+    defaults = parser.parse_args(args=args)
+    defaults = dict(defaults._get_kwargs(), **arg_options)
+    args = defaults.pop('args', ())
+    if 'skip_checks' not in options:
+        defaults['skip_checks'] = True
```
Comment (ta2-1) [inline tests/core/management/subcommands.py]: how is it used?

File: pootle/core/management/subcommands.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        if is_subcommand:
+            # if there is a subcommand then move it in argv to the front of other
+            # params and defer to the the subcommand
+            args = list(args)
+            del args[args.index(known.subcommand)]
+            parser = self.cmd.subcommands[
+                known.subcommand]().create_parser(self.prog, known.subcommand)
+            options = parser.parse_args(tuple(args))
+            options.subcommand = known.subcommand
+            options.super = self.cmd
```
Comment (ta2-1) [inline pootle/core/management/subcommands.py]: is it the best name for main command?

Comment (phlax) [inline pootle/core/management/subcommands.py]: im removing this - i have already done so elsewhere

File: tests/core/management/subcommands.py (hunk 6 above)
Comment (phlax) [inline tests/core/management/subcommands.py]: ill remove

File: tests/core/management/subcommands.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+            self.stdout.write("Did a foo")
+
+    foo_command = FooCommand()
+    assert not foo_command.subcommands
+    call_command_object(foo_command)
+    out, err = capsys.readouterr()
+    assert out == u'Did a foo\n'
+
+
+@pytest.fixture(params=["command", "argv"])
```
Comment (ta2-1) [inline tests/core/management/subcommands.py]: why did you put fixtures in tests module?

Comment (phlax) [inline tests/core/management/subcommands.py]: ill move now

**Comment by ta2-1** [comment]
lgtm

**Pull request merged**